### PR TITLE
Pubdev 3866 Stratified_split should be better than random sampling

### PR DESCRIPTION
**Pull request by mklechan**
current stratified_split does nothing more than a regular test_split. Added logic to have target distributions as close as possible to the original distribution

### CODE REVIEW

**Comment by mklechan** [comment]
@h2o-ops please test

**Comment by mklechan** [comment]
tested on multinode bigdata/server/airlines dataset on "Origin" column with 380 factors. On two servers takes ~5min

**Comment by navdeep-G** [comment]
@mklechan Might want to figure out why those other commits are in there. Also good to squash.

**Comment by navdeep-G** [comment]
@mklechan How long does the multinode test take on current master?

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java [hunk 1]
```diff
@@ -49,23 +57,47 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
     for (int i = 0; i < nClass; ++i)
       seeds[i] = getRNG(seed + i).nextLong();
     String[] dom = new String[]{"train", "test"};
-    return new ValFrame(new MRTask() {
-      private boolean isTest(int row, long seed) {
-        return getRNG(row + seed).nextDouble() <= testFrac;
-      }
+    // create frame with all 0s (default is train)
+    Key<Frame> k1 = Key.make();
```
Comment (navdeep-G) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java]: Maybe try: `Key<Frame> k1 = fr._key;`

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java [hunk 2]
```diff
@@ -49,23 +57,47 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
+    Frame result = new Frame(k1, new String[]{"test_train_split"}, new Vec[]{Vec.makeCon(0.0,fr.anyVec().length())});
+    // create index frame
+    Key<Frame> k2 = Key.make();
+    Frame ones = new Frame(k2, new String[]{"ones"}, new Vec[]{Vec.makeCon(1.0,fr.anyVec().length())});
+    DKV.put(ones);
+    System.out.println(ones._key);
```
Comment (navdeep-G) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java]: Is this print statement necessary?

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java [hunk 3]
```diff
@@ -49,23 +57,47 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
+    Frame idx = Rapids.exec("(cumsum " + ones._key + " 0)").getFrame();
+    DKV.put(idx);
+    System.out.println(idx._key);
```
Comment (navdeep-G) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java]: Is this print statement necessary?

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java [hunk 4]
```diff
@@ -49,23 +57,47 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
-          for (int row = 0; row < y._len; ++row) {
-            if (y.at8(row) == (classes == null ? classLabel : classes[classLabel])) {
-              if (isTest(start + row, seeds[classLabel])) ss.addNum(1, 0);
-              else ss.addNum(0, 0);
-            }
-          }
+        // extract frame with index locations of the minority class
+        int clabel = classes == null ? classLabel : (int) classes[classLabel];
+        Frame idxFound = Rapids.exec("(rows " + idx._key + " (== " + inputFrKey + " " + clabel + "))").getFrame();
+        System.out.println(idxFound._key);
```
Comment (navdeep-G) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java]: Is this print statement necessary?

File: h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java [hunk 5]
```diff
@@ -49,23 +57,47 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
+          // update the train/test frame
+          result.anyVec().set(idxFound.anyVec().at8(i) - 1, 1.0);
+          generated += 1;
+          count += 1;
         }
-      }
-    }.doAll(1, Vec.T_NUM, new Frame(y)).outputFrame(new String[]{"test_train_split"}, new String[][]{dom}));
+        idxFound.delete();
+    }
+
```
Comment (navdeep-G) [inline h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java]: Add a comment about cleaning up DKV

**Pull request closed without merging**